File: library/src/main/java/de/westnordost/countryboundaries/CountryBoundaries.java
package de.westnordost.countryboundaries;

import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class CountryBoundaries
{
	final CountryBoundariesCell[] raster;
	final int rasterWidth;
	final Map<String, Double> geometrySizes;

	CountryBoundaries(
			CountryBoundariesCell[] raster, int rasterWidth, Map<String, Double> geometrySizes)
	{
		this.raster = raster;
		this.rasterWidth = rasterWidth;
		this.geometrySizes = geometrySizes;
	}

	public static CountryBoundaries load(InputStream is) throws IOException
	{
		return new CountryBoundariesDeserializer().read(new DataInputStream(is));
	}

	/** @param longitude longitude of geo position (-180...180)
	 *  @param latitude latitude of geo position (-90...90)
	 *  @param ids ids of the countries to look for. Note that if you have many ids, you should use
	 *             a Set to increase performance
	 *
	 *  @throws IllegalArgumentException if any parameter is not finite (NaN or Infinite)
	 *          or latitude is not between -90.0 and +90.0
	 *
	 *  @return whether the given position is in any of the countries with the given ids */
	public boolean isInAny(double longitude, double latitude, Collection<String> ids)
	{
		validatePosition(longitude, latitude);
		longitude = normalize(longitude, -180.0, 360.0);
		int cellX = longitudeToCellX(longitude);
		int cellY = latitudeToCellY(latitude);
		int localX = longitudeToLocalX(cellX, longitude);
		int localY = latitudeToLocalY(cellY, latitude);
		return getCell(cellX, cellY).isInAny(localX, localY, ids);
	}

	/** @param longitude longitude of geo position (-180...180)
	 *  @param latitude latitude of geo position (-90...90)
	 *  @param id id of the country to look for
	 *
	 *  @throws IllegalArgumentException if any parameter is not finite (NaN or Infinite)
	 *          or latitude is not between -90.0 and +90.0
	 *
	 *  @return whether the given position is in the country with the given id */
	public boolean isIn(double longitude, double latitude, String id)
	{
		return isInAny(longitude, latitude, Collections.singleton(id));
	}

	/** @param longitude longitude of geo position (-180...180)
	 *  @param latitude latitude of geo position (-90...90)
	 *  @return the ids of the countries the given position is contained in, ordered by size ascending */
	public List<String> getIds(double longitude, double latitude)
	{
		validatePosition(longitude, latitude);
		longitude = normalize(longitude, -180.0, 360.0);
		int cellX = longitudeToCellX(longitude);
		int cellY = latitudeToCellY(latitude);
		int localX = longitudeToLocalX(cellX, longitude);
		int localY = latitudeToLocalY(cellY, latitude);

		List<String> result = getCell(cellX, cellY).getIds(localX, localY);
		Collections.sort(result, this::compareSize);
		return result;
	}

	/** Identify which countries are guaranteed to contain the given bounding box fully.
	 *  The given bounding box may wrap around the 180th longitude, i.e minLongitude = 170 and
	 *  maxLongitude = -170.
	 *
	 *  @param minLongitude minimum longitude of geo position (-180...180)
	 *  @param minLatitude minimum latitude of geo position (-90...90)
	 *  @param maxLongitude maximum longitude of geo position (-180...180)
	 *  @param maxLatitude maximum latitude of geo position (-90...90)
	 *
	 *  @throws IllegalArgumentException if any parameter is not finite (NaN or Infinite),
	 *          minLatitude is greater than maxLatitude or any latitude is not between
	 *          -90.0 and +90.0
	 *
	 *  @return the ids of the countries the given bounding box is guaranteed to be contained in,
	 *          not in any particular order */
	public Set<String> getContainingIds(
			double minLongitude, double minLatitude, double maxLongitude, double maxLatitude)
	{
		Set<String> ids = new HashSet<>();
		final boolean[] firstCell = {true};
		forCellsIn(minLongitude, minLatitude, maxLongitude, maxLatitude, cell ->
		{
			if(firstCell[0])
			{
				ids.addAll(cell.containingIds);
				firstCell[0] = false;
			}
			else
			{
				ids.retainAll(cell.containingIds);
				if (ids.isEmpty()) return false;
			}
			return true;
		});
		return ids;
	}

	/** Identify which countries intersect with the given bounding box. The given bounding box may
	 *  wrap around the 180th longitude, i.e minLongitude = 170 and maxLongitude = -170.
	 *
	 *  @param minLongitude minimum longitude of geo position (-180...180)
	 *  @param minLatitude minimum latitude of geo position (-90...90)
	 *  @param maxLongitude maximum longitude of geo position (-180...180)
	 *  @param maxLatitude maximum latitude of geo position (-90...90)
	 *
	 *  @throws IllegalArgumentException if any parameter is not finite (NaN or Infinite),
	 *          minLatitude is greater than maxLatitude or any latitude is not between
	 *          -90.0 and +90.0
	 *
	 *  @return the ids of the countries the given bounding box intersects with, not in any
	 *          particular order */
	public Set<String> getIntersectingIds(
			double minLongitude, double minLatitude, double maxLongitude, double maxLatitude)
	{
		Set<String> ids = new HashSet<>();
		forCellsIn(minLongitude, minLatitude, maxLongitude, maxLatitude, cell ->
		{
			ids.addAll(cell.getAllIds());
			return true;
		});
		return ids;
	}

	private void forCellsIn(
			double minLongitude, double minLatitude, double maxLongitude, double maxLatitude,
			CellRunnable runnable)
	{
		if (!Double.isFinite(minLongitude))
			throw new IllegalArgumentException("minLongitude must be finite");
		if (!Double.isFinite(minLatitude))
			throw new IllegalArgumentException("minLatitude must be finite");
		if (!Double.isFinite(maxLongitude))
			throw new IllegalArgumentException("maxLongitude must be finite");
		if (!Double.isFinite(maxLatitude))
			throw new IllegalArgumentException("maxLatitude must be finite");
		if (minLatitude < -90 || minLatitude > 90)
			throw new IllegalArgumentException("minLatitude is out of bounds");
		if (maxLatitude < -90 || maxLatitude > 90)
			throw new IllegalArgumentException("maxLatitude is out of bounds");
		if (minLatitude > maxLatitude)
			throw new IllegalArgumentException("maxLatitude is smaller than minLatitude");

		minLongitude = normalize(minLongitude, -180, 360);
		maxLongitude = normalize(maxLongitude, -180, 360);
		int minX = longitudeToCellX(minLongitude);
		int maxY = latitudeToCellY(minLatitude);
		int maxX = longitudeToCellX(maxLongitude);
		int minY = latitudeToCellY(maxLatitude);

		// might wrap around
		int stepsX = minX > maxX ? rasterWidth - minX + maxX : maxX - minX;

		for (int xStep = 0; xStep <= stepsX; ++xStep)
		{
			int x = (minX + xStep) % rasterWidth;
			for (int y = minY; y <= maxY; y++)
			{
				if (!runnable.run(getCell(x, y))) return;
			}
		}
	}

	private interface CellRunnable
	{
		boolean run(CountryBoundariesCell cell);
	}

	private CountryBoundariesCell getCell(int x, int y) {
		return raster[y * rasterWidth + x];
	}

	private void validatePosition(double longitude, double latitude) {
		if (!Double.isFinite(longitude))
			throw new IllegalArgumentException("longitude must be finite");
		if (!Double.isFinite(latitude))
			throw new IllegalArgumentException("latitude must be finite");
		if (latitude < -90 || latitude > 90)
			throw new IllegalArgumentException("latitude is out of bounds");
	}

	private int longitudeToCellX(double longitude) {
		return (int) Math.min(
				rasterWidth - 1,
				Math.floor(rasterWidth * (180 + longitude) / 360.0)
		);
	}

	private int latitudeToCellY(double latitude) {
		int rasterHeight = raster.length / rasterWidth;
		return (int) Math.max(
				0,
				Math.ceil(rasterHeight * (90 - latitude) / 180.0) - 1
		);
	}

	private int longitudeToLocalX(int cellX, double longitude) {
		double cellLongitude = -180.0 + 360.0 * cellX / rasterWidth;
		return (int) ((longitude - cellLongitude) * 360.0 * 0xffff / rasterWidth);
	}

	private int latitudeToLocalY(int cellY, double latitude) {
		int rasterHeight = raster.length / rasterWidth;
		double cellLatitude = +90 - 180.0 * (cellY + 1) / rasterHeight;
		return (int) ((latitude - cellLatitude) * 180.0 * 0xffff / rasterHeight);
	}

	private static double normalize(double value, double startAt, double base)
	{
		value = value % base;
		if (value < startAt) value += base;
		else if (value >= (startAt + base)) value -= base;
		return value;
	}

	private double getSize(String id)
	{
		Double size = geometrySizes.get(id);
		return size != null ? size : 0;
	}

	private int compareSize(String isoCode1, String isoCode2)
	{
		return (int) (getSize(isoCode1) - getSize(isoCode2));
	}

	@Override public boolean equals(Object o)
	{
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		CountryBoundaries that = (CountryBoundaries) o;

		return
				rasterWidth == that.rasterWidth
						&& Arrays.equals(raster, that.raster)
						&& geometrySizes.equals(that.geometrySizes);
	}

	@Override public int hashCode()
	{
		return rasterWidth + 31 * (Arrays.hashCode(raster) + 31 * geometrySizes.hashCode());
	}
}

File: library/src/commonMain/kotlin/de/westnordost/countryboundaries/CountryBoundaries.kt
package de.westnordost.countryboundaries

import kotlinx.io.Sink
import kotlinx.io.Source
import kotlin.math.ceil
import kotlin.math.floor

/**
 * A spatial index to look up in which country a given geo position is located fast.
 *
 * Construct with [CountryBoundaries.Companion.deserializeFrom].
 * */
public class CountryBoundaries internal constructor(
    internal val raster: List<CountryBoundariesCell>,
    internal val rasterWidth: Int,
    internal val geometrySizes: Map<String, Double>
) {
    private val rasterHeight = raster.size / rasterWidth

    /**
     * @param longitude longitude of geo position (-180...180)
     * @param latitude latitude of geo position (-90...90)
     * @param ids ids of the countries to look for. Note that if you have many ids, you should use
     *            a Set to increase performance
     *
     * @throws IllegalArgumentException if any parameter is not finite (NaN or Infinite) or
     *                                  latitude is not between -90.0 and +90.0
     *
     * @return whether the given position is in any of the countries with the given ids
     */
    public fun isInAny(longitude: Double, latitude: Double, ids: Collection<String>): Boolean {
        var longitude = longitude
        validatePosition(longitude, latitude)
        longitude = normalizeLongitude(longitude)
        val cellX = longitudeToCellX(longitude)
        val cellY = latitudeToCellY(latitude)
        val localX = longitudeToLocalX(cellX, longitude)
        val localY = latitudeToLocalY(cellY, latitude)

        return getCell(cellX, cellY).isInAny(Point(localX, localY), ids)
    }

    /**
     * @param longitude longitude of geo position (-180...180)
     * @param latitude latitude of geo position (-90...90)
     * @param id id of the country to look for
     *
     * @throws IllegalArgumentException if any parameter is not finite (NaN or Infinite)
     *         or latitude is not between -90.0 and +90.0
     *
     * @return whether the given position is in the country with the given id
     */
    public fun isIn(longitude: Double, latitude: Double, id: String): Boolean {
        return isInAny(longitude, latitude, listOf(id))
    }

    /**
     * @param longitude longitude of geo position (-180...180)
     * @param latitude latitude of geo position (-90...90)
     * @return the ids of the countries the given position is contained in, ordered by size
     *         ascending
     */
    public fun getIds(longitude: Double, latitude: Double): List<String> {
        var longitude = longitude
        validatePosition(longitude, latitude)
        longitude = normalizeLongitude(longitude)
        val cellX = longitudeToCellX(longitude)
        val cellY = latitudeToCellY(latitude)
        val localX = longitudeToLocalX(cellX, longitude)
        val localY = latitudeToLocalY(cellY, latitude)

        return getCell(cellX, cellY)
            .getIds(Point(localX, localY))
            .sortedBy { geometrySizes[it] ?: 0.0 }
    }

    /**
     * Identify which countries are guaranteed to contain the given bounding box fully.
     * The given bounding box may wrap around the 180th longitude, e.g. minLongitude = 170 and
     * maxLongitude = -170.
     *
     * @param minLongitude minimum longitude of geo position (-180...180)
     * @param minLatitude minimum latitude of geo position (-90...90)
     * @param maxLongitude maximum longitude of geo position (-180...180)
     * @param maxLatitude maximum latitude of geo position (-90...90)
     *
     * @throws IllegalArgumentException if any parameter is not finite (NaN or Infinite),
     *                                  minLatitude is greater than maxLatitude or any latitude is
     *                                  not between -90.0 and +90.0
     *
     * @return the ids of the countries the given bounding box is guaranteed to be contained in,
     *         not in any particular order
     */
    public fun getContainingIds(
        minLongitude: Double, minLatitude: Double, maxLongitude: Double, maxLatitude: Double
    ): Set<String> {
        val ids = HashSet<String>()
        var firstCell = true
        for (cell in cellsIn(minLongitude, minLatitude, maxLongitude, maxLatitude)) {
            if (firstCell) {
                ids.addAll(cell.containingIds)
                firstCell = false
            } else {
                ids.retainAll(cell.containingIds)
                if (ids.isEmpty()) break
            }
        }
        return ids
    }

    /**
     * Identify which countries may intersect with the given bounding box. In other words, any
     * point within the given bounding box can only be in any of the returned countries (or none).
     *
     * So, from the technical point of view, it just returns which countries are in the same
     * cell(s) of the raster as the bounding box.
     *
     * The given bounding box may
     * wrap around the 180th longitude, i.e minLongitude = 170 and maxLongitude = -170.
     *
     * @param minLongitude minimum longitude of geo position (-180...180)
     * @param minLatitude minimum latitude of geo position (-90...90)
     * @param maxLongitude maximum longitude of geo position (-180...180)
     * @param maxLatitude maximum latitude of geo position (-90...90)
     *
     * @throws IllegalArgumentException if any parameter is not finite (NaN or Infinite),
     * minLatitude is greater than maxLatitude or any latitude is not between
     * -90.0 and +90.0
     *
     * @return the ids of the countries the given bounding box may intersects with, not in any
     * particular order
     */
    public fun getIntersectingIds(
        minLongitude: Double, minLatitude: Double, maxLongitude: Double, maxLatitude: Double
    ): Set<String> {
        val ids = HashSet<String>()
        for (cell in cellsIn(minLongitude, minLatitude, maxLongitude, maxLatitude)) {
            ids.addAll(cell.getAllIds())
        }
        return ids
    }

    private fun cellsIn(
        minLongitude: Double, minLatitude: Double, maxLongitude: Double, maxLatitude: Double,
    ): Sequence<CountryBoundariesCell> = sequence {
        var minLongitude = minLongitude
        var maxLongitude = maxLongitude
        validateBounds(minLongitude, minLatitude, maxLongitude, maxLatitude)
        minLongitude = normalizeLongitude(minLongitude)
        maxLongitude = normalizeLongitude(maxLongitude)
        val minX = longitudeToCellX(minLongitude)
        val maxY = latitudeToCellY(minLatitude)
        val maxX = longitudeToCellX(maxLongitude)
        val minY = latitudeToCellY(maxLatitude)

        // might wrap around
        val stepsX = if (minX > maxX) rasterWidth - minX + maxX else maxX - minX

        for (xStep in 0..stepsX) {
            val x = (minX + xStep) % rasterWidth
            for (y in minY..maxY) {
                yield(getCell(x, y))
            }
        }
    }

    private fun getCell(x: Int, y: Int): CountryBoundariesCell =
        raster[y * rasterWidth + x]

    private fun longitudeToCellX(longitude: Double): Int =
        floor(rasterWidth * (180 + longitude) / 360.0).toInt().coerceAtMost(rasterWidth - 1)

    private fun latitudeToCellY(latitude: Double): Int =
        (ceil(rasterHeight * (90 - latitude) / 180.0) - 1).toInt().coerceAtLeast(0)

    private fun longitudeToLocalX(cellX: Int, longitude: Double): UShort {
        val cellLongitude = -180.0 + 360.0 * cellX / rasterWidth
        return ((longitude - cellLongitude) * rasterWidth * 0xffff / 360.0).toInt().toUShort()
    }

    private fun latitudeToLocalY(cellY: Int, latitude: Double): UShort {
        val cellLatitude = +90 - 180.0 * (cellY + 1) / rasterHeight
        return ((latitude - cellLatitude) * rasterHeight * 0xffff / 180.0).toInt().toUShort()
    }

    private fun validateBounds(
        minLongitude: Double, minLatitude: Double, maxLongitude: Double, maxLatitude: Double
    ) {
        require(minLongitude.isFinite()) { "minLongitude must be finite" }
        require(minLatitude.isFinite()) { "minLatitude must be finite" }
        require(maxLongitude.isFinite()) { "maxLongitude must be finite" }
        require(maxLatitude.isFinite()) { "maxLatitude must be finite" }
        require(minLatitude >= -90 && minLatitude <= 90) { "minLatitude is out of bounds" }
        require(maxLatitude >= -90 && maxLatitude <= 90) { "maxLatitude is out of bounds" }
        require(minLatitude <= maxLatitude) { "maxLatitude is smaller than minLatitude" }
    }

    private fun validatePosition(longitude: Double, latitude: Double) {
        require(longitude.isFinite()) { "longitude must be finite" }
        require(latitude.isFinite()) { "latitude must be finite" }
        require(latitude >= -90 && latitude <= 90) { "latitude is out of bounds" }
    }

    private fun normalizeLongitude(longitude: Double): Double =
        normalize(longitude, -180.0, 360.0)

    private fun normalize(value: Double, startAt: Double, base: Double): Double {
        var value = value
        value = value % base
        if (value < startAt) value += base
        else if (value >= (startAt + base)) value -= base
        return value
    }

    /** Serialize this RegionIndex to the given sink */
    internal fun serializeTo(sink: Sink) {
        sink.writeRegionIndex(this)
    }

    public companion object {
        /** Create a new RegionIndex by deserializing from the given source */
        public fun deserializeFrom(source: Source): CountryBoundaries =
            source.readRegionIndex()
    }
}